i2c - allow shell commands to operate on up to 32-bit registers

barnas-michal opened the pull request:
This PR add support to i2c shell commands for up to 32-bit addressed registers in i2c. Some chips may have more than 256 registers and needs 16-bit support to fully cover them.
Also it changes the functions parameter name to meet the MISRA requirements.

barnas-michal committed:
i2c: allow shell commands to operate on up to 32-bit registers

Improve shell commands to support up to 32-bit addresses.
The check if function should use 8/16/24/32-bit address is basing
on the length of address used in function parameter. To address
registers below 256 using 16-bit, one should write them left-padded
with zeroes. Address is sent in big endian.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

teburd reviewed (changes requested):
First of all thank you for your contribution. I noted some minor nits but please read the rest here first. The bug fixes could and should be a PR on their own! Those are always much appreciated.

I'm not a fan of expanding the I2C API here to be frank. There's many different ways a device may do register addressing, all of them can be covered by i2c_write_read wrapped by a particular device driver. That seems like the right place to deal with register addressing scheme concerns. I2C doesn't intrinsically understand or care about register addressing schemes.

You could point at the existing calls that deal with register addresses here, and you aren't wrong to do so, but that has in fact bitten this API once already in i2c_burst_write.

Adding to the shell commands I think could be sensible with some ifdefry and a Kconfig option for it as I'd guess the majority of folks aren't interested. I can see it being potentially useful in some poking around. Those two read/write calls could be wrappers around i2c_write_read as would be expected.

Perhaps a more generalized i2c write_read shell command would be the way to go though instead as it could handle all these cases much better. 

There are after all several schemes to go beyond the very common 8 bit register addressing scheme many i2c devices do. Whats the address byte order here? Is it always the same on all devices? Is there a 8bit register denoting a register page for several pages of 8 bit register addresses?

semihalf-barnas-michal commented on drivers/i2c/i2c_shell.c:
```diff
@@ -130,7 +131,13 @@ static int cmd_i2c_write(const struct shell *shell, size_t argc, char **argv)
+		ret = i2c_reg16_burst_write(dev, dev_addr, reg_addr,
+					    buf, num_bytes);
+	else
+		ret = i2c_burst_write(dev, dev_addr, reg_addr, buf, num_bytes);
```
I have changed this PR, removing the API change and improving only shell. Burst was removed, hope that it will be nice change :)

semihalf-barnas-michal commented on drivers/i2c/i2c_shell.c:
```diff
@@ -218,7 +238,13 @@ static int cmd_i2c_read(const struct shell *shell, size_t argc, char **argv)
+		ret = i2c_reg16_burst_read(dev, dev_addr, reg_addr,
+					   buf, num_bytes);
+	else
+		ret = i2c_burst_read(dev, dev_addr, reg_addr, buf, num_bytes);
```
Done

teburd commented on drivers/i2c/i2c_shell.c:
```diff
@@ -103,65 +103,108 @@ static int cmd_i2c_recover(const struct shell *shell,
 	int reg_addr;
 	int dev_addr;
-	int out_byte;
+	int ret;
```
This seems unused at the moment?

semihalf-barnas-michal replied:
I am assigning the result of function to check it in the if statement. It's only used in if, but the command call is too long to wrap it directly with if

semihalf-barnas-michal commented:
> First of all thank you for your contribution. I noted some minor nits but please read the rest here first. The bug fixes could and should be a PR on their own! Those are always much appreciated.

Thank you very much for this review. I really appreciate that you explained your viewpoint and also tried to understand mine and explained it :) I have divided the fixes and they will be sent as individual PR.

> I'm not a fan of expanding the I2C API here to be frank. There's many different ways a device may do register addressing, all of them can be covered by i2c_write_read wrapped by a particular device driver. That seems like the right place to deal with register addressing scheme concerns. I2C doesn't intrinsically understand or care about register addressing schemes.
> You could point at the existing calls that deal with register addresses here, and you aren't wrong to do so, but that has in fact bitten this API once already in i2c_burst_write.

You are right. I was worried about expanding the header file too much, and you concluded my worries perfectly.
 
> Adding to the shell commands I think could be sensible with some ifdefry and a Kconfig option for it as I'd guess the majority of folks aren't interested. I can see it being potentially useful in some poking around. Those two read/write calls could be wrappers around i2c_write_read as would be expected.
> Perhaps a more generalized i2c write_read shell command would be the way to go though instead as it could handle all these cases much better.

I have removed the API change and extracted the shell logic for write/read and write_byte/read_byte. Also removed the usage of burst command and added the support for higher-bits addresses.

> There are after all several schemes to go beyond the very common 8 bit register addressing scheme many i2c devices do. Whats the address byte order here? Is it always the same on all devices? Is there a 8bit register denoting a register page for several pages of 8 bit register addresses?

As I have tried to find, most informations I have found were about big endian, so that's what I used. However I understand your statement, as it is like "de facto" standard, but no one guarantees that there are no little-endian chips.
I tried to refactor the shell functions, so I hope that now the API is not changed, nor expanded, but still adds additional support for higher bits addresses.

teburd reviewed (changes requested):
Almost there, this seems really nice. I like the combined shell commands with addresses, really you could always take the order the  bytes are written in the shell as the byte order to use so, very nice I think. I perhaps wouldn't bother even doing any byte ordering here, let the commands from the shell determine that. Now you support both byte orders.

 The unused ret and maybe a fix for the strtol + strlen combo for determining reg address bytes is needed if I read it correctly. 

Given those minor things are corrected, a happy +1 to this PR from me

nordic-krch commented:
Please update PR title and description.

barnas-michal committed:
i2c: fix for MISRA in i2c shell commands

This commit changes the shell parameter name to meet the MISRA
check requirements.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

semihalf-barnas-michal commented on drivers/i2c/i2c_shell.c:
```diff
@@ -103,65 +103,108 @@ static int cmd_i2c_recover(const struct shell *shell,
 
-	if (i2c_reg_write_byte(dev, dev_addr, reg_addr, out_byte) < 0) {
-		shell_error(shell, "Failed to write to device: %s", argv[1]);
+	reg_addr_bytes = MIN(4, (strlen(s_reg_addr) + 1) / 2);
```
Right, I forgot about 0x prefix. I've added function that takes it into account (based on strtol implementation) and are used in both read and write now. Also empty "0x" will be used as 1-byte 0 address.

semihalf-barnas-michal commented:
> Please update PR title and description.

Done, thanks.

MaureenHelm commented:
After this PR is accepted into main, please submit an additional PR to backport the MISRA fixes to `v2.7-auditable-branch`. Unfortunately we can't use the automatic backport bot here because this PR contains additional changes that shouldn't be backported.

semihalf-barnas-michal commented:
> After this PR is accepted into main, please submit an additional PR to backport the MISRA fixes to `v2.7-auditable-branch`. Unfortunately we can't use the automatic backport bot here because this PR contains additional changes that shouldn't be backported.

Sure, I will submit it.

jakub-uC commented:
@semihalf-barnas-michal : Do you think you can replace the magic number with some defines?

semihalf-barnas-michal commented:
@jakub-uC : sure, please check if this is what you expected.

nordic-krch reviewed (approved):
looks good

mbolivar-nordic commented on drivers/i2c/i2c_shell.c:
```diff
@@ -88,82 +103,125 @@ static int cmd_i2c_recover(const struct shell *shell_ctx,
 	if (!dev) {
-		shell_error(shell_ctx, "I2C: Device driver %s not found.", argv[1]);
+		shell_error(shell_ctx, "I2C: Device driver %s not found.",
+			    argv[1]);
```
Shouldn't this be `ARGV_DEV` instead of `1`, since this patch is getting rid of the magic number?

semihalf-barnas-michal replied:
I've started refactoring this, but it was merged in the meantime. Should I revert that change and create new PR with changes and improvements, or just add new PR only with improvements?

mbolivar-nordic reviewed (commented):
The last patch is mixing up cosmetic improvements and features in a way that is hard to review; it'd be nice to see that cleaned up before merge. I noticed one issue with what seems like a magic number cleanup in the patch that is adding a feature, but it's a bit hard to tell what's going on from a quick skim, so I'm not going to merge this now.
